### DISCUSSION TITLE
findUnique behavior

### DISCUSSION THREAD

**Discussion by thejoyfuldev** [prisma/prisma#12642] [Q&A] [answered]
Team,

Just started fiddling with Prisma and so far enjoy the ease of it.  One question I do have is the ability to findUnique when dealing with user records.  Specifically, during login the intent is to find a unique record that matches against the email and password.  But, unless I am missing something, the where option only allows parameters that were marked as ```@unique``` in the schema.

So, attempting to do:

```
  const user = await db.user.findUnique({
    where: {
      email: {
        equals: email,
      },
      password: {
        equals: passwordHash,
      },
    },
  });
```

creates a typescript error:

```Type '{ equals: string; }' is not assignable to type 'string'.``` against the email parameter.

I've also tried to do different variants such as 
```
email: email,
password: passwordHash
```

but that throws an error ```Type '{ email: string; password: string; }' is not assignable to type 'UserWhereUniqueInput'```

However, what does work is changing from findUnique to findFirst or findMany but to me the intent of findUnique (previously findOne) is lost.

Works with no issues:
```
  const user = await db.user.findFirst({
    where: {
      email: email,
      password: passwordHash,
    },
  });
```

But again, while I am finding the first record, I'm really wanting the unique or one record.  Maybe it's just naming convention where findOne isn't the same as findUnique and findFirst doesn't necessarily correlate with findOne although that's what is wanted in this case.

**Comment by austincrim** [comment, marked as answer]
Hey @jasonabullard 👋, welcome to the Prisma community! I'm glad you are enjoying it so far 😄.

> the where option only allows parameters that were marked as @unique in the schema.

This is correct! `findUnique` can only be guaranteed to return a unique row if you are filtering by a field that is actually unique across the table. In the case of a user account, the `email` field is generally required to be unique, therefore annotating it with the `@unique` attribute in your Prisma schema makes sense and would enable the use case you outlined above.

If a combination of multiple fields is guaranteed to be unique, you could use a compound unique constraint like so: `@@unique([email, password])`.

[Relevant docs page.](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#defining-a-unique-field)

Let me know if I can clarify further!

**Comment by Prashant-S29** [comment]
Hey @austincrim,
I just started working on Prisma, and faced the same issue. For me it goes like this

![image](https://github.com/prisma/prisma/assets/112539218/ebd09921-0ae8-4f60-9492-5e32b9566c1e)

This is my code for signup page. API working all good, tested it with postman and worked as expected. But when deploying to vercel, I got this error message 
`
Type error: Object literal may only specify known properties, and 'userEmail' does not exist in type 'UserWhereInput'.
--     // check if userEmail already exists
  const existingUserByEmail = await prisma_db.user.findFirst({
      where: { userEmail: userEmail },
                                         ^
`

![image](https://github.com/prisma/prisma/assets/112539218/fc1b94f7-567a-4c6c-9635-a60ba22f4251)

I have also tried replacing findUnique to findOne, but it didn't work

This is my Schema 
![image](https://github.com/prisma/prisma/assets/112539218/6a4b8caf-d1af-4d21-af46-2b07aa7c1a6a)


can you pls suggest some solutions

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
